### PR TITLE
root_scalar: a new method that improves upon the Newton's method 

### PR DESCRIPTION
**Pull request by ankushaggarwal**
#### What does this implement/fix?

An "extended-Newton" method is implemented in root_scalar function based on https://doi.org/10.3390/a13040078 

This extension improves the domain of convergence, especially for non Lipschitz functions without any extra computation required (still uses only one function and first derivative evaluation per iteration). The only extra is a second guess x1, which if not supplied is assigned as a perturbation of x0. The new method can also be seen as a combination of Newton and secant methods, and somewhere between Newton's and Halley's methods.

#### Additional information

I have tried to maintain the backward compatability. The default options remain the same and the standard Newton continues to work as is. I have also added the necessary documentation for the new method.

### CODE REVIEW

**Comment by rgommers** [comment]
@ankushaggarwal thanks for proposing your method as an addition to our collection of root finders.

The paper doesn't seem to have any citations, and there's no benchmarks here either that show this is clearly an improvement on what we already have. http://scipy.github.io/devdocs/dev/core-dev/index.html#deciding-on-new-features gives some guidance on how we'd decide on this. For something as fundamental as root finding, I'd expect 50-100+ citations and/or convincing benchmarks. Do you have benchmarks (accuracy, not speed)?

gh-12620 may be related. I'll also add @ev-br's comment to that on the mailing list, which is relevant here:

_FWIW I'd always recommend brentq for scalar root-finding._

_Trying to make newton-based methods more robust seems to be more or less reinventing that wheel, which has already been implemented (don't know who implemented it; hope you read this thread so that you see my thank you!) and is as robust  as it gets._

**Comment by ankushaggarwal** [comment]
@rgommers Thanks for your comment. This is my first time contributing to scipy. How can I add benchmarks? And what do you mean accuracy, not speed?

**Comment by rgommers** [comment]
> How can I add benchmarks? And what do you mean accuracy, not speed?

A set of Python functions on which you run root finding. And then seeing how many problems get solved with the existing methods in `scipy.optimize` and with your proposed new method.

**Comment by ankushaggarwal** [comment]
> > How can I add benchmarks? And what do you mean accuracy, not speed?
> 
> A set of Python functions on which you run root finding. And then seeing how many problems get solved with the existing methods in `scipy.optimize` and with your proposed new method.

OK thanks! I will put that together. I suppose that will not go into the repository, and I can just share a script here in the comments?

**Comment by rgommers** [comment]
> I suppose that will not go into the repository, and I can just share a script here in the comments?

You can just start with a script. We may want to turn it into regression tests or benchmarks (see `benchmarks/benchmarks/optimize_zeros.py`) afterwards.

File: scipy/optimize/__init__.py [hunk 1]
```diff
@@ -195,6 +196,8 @@
 +-------------+----------+----------+-----------+-------------+-------------+----------------+
 | `R` or `C`  | No       | Yes      | No        | newton      | No          | 2.00 (1.41)    |
 +-------------+----------+----------+-----------+-------------+-------------+----------------+
+| `R` or `C`  | No       | Yes      | No        | exnewton    | No	    | >2.00 (1.41)   |
++-------------+----------+----------+-----------+-------------+-------------+----------------+
```
Comment (pvanmulbregt) [inline scipy/optimize/__init__.py:200]: No need for the line of "-" as the column values for Bracket/fprime/fprime2/Solvers are the same as for `newton`. Similar to the collection of bisect/brentq/.../toms748

**Comment by pvanmulbregt** [comment]
Regression testing for `exnewton` should be added to `optimize/tests/test_zeros.py`.  You can easily add `test_exnewton_collections()` by just cloning `test_newton_collections`, replacing `newton` with `exnewton`, and possibly setting `known_fail`.

**Comment by mdhaber** [comment]
@ankushaggarwal thanks for the PR. Were you interested in finishing this up?

**Comment by ankushaggarwal** [comment]
Yes, I am interested in finishing this up. Really sorry, I got distracted with other things and lost track of this one. I will work on getting the benchmark tests and get back.

**Comment by mdhaber** [comment]
Ok! Let me know if you'd like this to be re-opened once you've merged main and developed benchmarks.

**Pull request closed without merging**